### ISSUE TITLE
Padding difference in Stable Diffusion tokenizers: 1.x vs 2.x vs OpenCLIP

### ISSUE THREAD

**Issue by jorgemcgomes** [huggingface/diffusers#2201]
### Describe the bug

There is a difference between the tokenizers in SD 1.x and 2.x, which seems to be introduced by a typo/bug. The consequence is that 1.x is padding with EOS (49407), while 2.x is padding with ! (0).

I know these two models use different CLIP models, but the tokenizer should be the same.
If we check the tokenizer used by 2.x (OpenCLIP ViT-H-14), padding should indeed be done with EOS, not 0:

https://huggingface.co/runwayml/stable-diffusion-v1-5/blob/39593d5650112b4cc580433f6b0435385882d819/tokenizer/special_tokens_map.json#L16

https://huggingface.co/stabilityai/stable-diffusion-2-base/blob/d28fc8045793886e512c5389771d3b3d560f9575/tokenizer/special_tokens_map.json#L16

https://huggingface.co/laion/CLIP-ViT-H-14-laion2B-s32B-b79K/blob/babb4f6a96d549f88b005dca40764a4274e914f6/special_tokens_map.json#L1

As shown below, the differences in embeddings are quite significant in the padding tokens, as expected.

And there are also significant differences in the generated images.

Generated with the built-in (bugged?) tokenizer (padding with zeros)
![image](https://user-images.githubusercontent.com/3987574/216054718-56922379-0863-45b6-b006-3f4816f76278.png)

Generated with the OpenCLIP tokenizer (padding with EOS)
![image](https://user-images.githubusercontent.com/3987574/216054879-c19e5c85-2fda-43a8-9a85-c35a1ac4b5c3.png)

Has diffusers SD 2.x been tested against the Stability AI implementation?
Is this really bug? Or a very very odd design choice?

The conversion script at https://github.com/huggingface/diffusers/blob/8267c7844504b55366525169187767ef92d1f499/src/diffusers/pipelines/stable_diffusion/convert_from_ckpt.py#L963 is using the v2 tokenizer, which is potentially bugged.

### Reproduction

https://colab.research.google.com/drive/1HNUPX7tA1P_ldH7fgCiclP1MtAONF51j?usp=sharing

```py
from transformers import AutoTokenizer

tokenizer_v1 = AutoTokenizer.from_pretrained("runwayml/stable-diffusion-v1-5", subfolder="tokenizer")
tokenizer_v2 = AutoTokenizer.from_pretrained("stabilityai/stable-diffusion-2-1-base", subfolder="tokenizer")
tokenizer_openclip = AutoTokenizer.from_pretrained("laion/CLIP-ViT-H-14-laion2B-s32B-b79K")
# the tokenizer call uses the same params as the stable diffusion pipelines
tokenizer_kwargs = dict(padding="max_length", max_length=tokenizer_v1.model_max_length, truncation=True, return_tensors="pt")

tok_v1 = tokenizer_v1("test tokenization", **tokenizer_kwargs)
tok_v2 = tokenizer_v2("test tokenization", **tokenizer_kwargs)      
tok_open = tokenizer_openclip("test tokenization", **tokenizer_kwargs)

tok_v1.input_ids
>>> tensor([[49406,  1628, 32634, 10847, 49407, 49407, 49407, 49407, 49407, 49407,
         49407, 49407, 49407, 49407, 49407, 49407, 49407, 49407, 49407, 49407,
         49407, 49407, 49407, 49407, 49407, 49407, 49407, 49407, 49407, 49407,
         49407, 49407, 49407, 49407, 49407, 49407, 49407, 49407, 49407, 49407,
         49407, 49407, 49407, 49407, 49407, 49407, 49407, 49407, 49407, 49407,
         49407, 49407, 49407, 49407, 49407, 49407, 49407, 49407, 49407, 49407,
         49407, 49407, 49407, 49407, 49407, 49407, 49407, 49407, 49407, 49407,
         49407, 49407, 49407, 49407, 49407, 49407, 49407]])

tok_v2.input_ids
>>> tensor([[49406,  1628, 32634, 10847, 49407,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0]])

tok_open.input_ids
>>> tensor([[49406,  1628, 32634, 10847, 49407, 49407, 49407, 49407, 49407, 49407,
         49407, 49407, 49407, 49407, 49407, 49407, 49407, 49407, 49407, 49407,
         49407, 49407, 49407, 49407, 49407, 49407, 49407, 49407, 49407, 49407,
         49407, 49407, 49407, 49407, 49407, 49407, 49407, 49407, 49407, 49407,
         49407, 49407, 49407, 49407, 49407, 49407, 49407, 49407, 49407, 49407,
         49407, 49407, 49407, 49407, 49407, 49407, 49407, 49407, 49407, 49407,
         49407, 49407, 49407, 49407, 49407, 49407, 49407, 49407, 49407, 49407,
         49407, 49407, 49407, 49407, 49407, 49407, 49407]])


from transformers import CLIPTextModel
model = CLIPTextModel.from_pretrained("stabilityai/stable-diffusion-2-1-base", subfolder="text_encoder")
embeddings_truth = model(tok_open.input_ids)[0]
embeddings_bugged = model(tok_v2.input_ids)[0]
diffs = (embeddings_truth - embeddings_bugged).abs()

# differences in the non-padding tokens
diffs[:, :5, :].max()
>>> 0

# differences in the padding tokens
diffs[:, 5:, :].max()
>>> 9.4606
```

Testing differences in images

```py
from diffusers import StableDiffusionPipeline
import torch

pipe = StableDiffusionPipeline.from_pretrained("stabilityai/stable-diffusion-2-1-base", revision="fp16", torch_dtype=torch.float16)
pipe.to("cuda")

generator = torch.Generator(device="cuda")
generator = generator.manual_seed(38)
img = pipe("An astronaut riding a dinossaur", generator=generator)[0]
img[0]

pipe.tokenizer = tokenizer_openclip
generator = torch.Generator(device="cuda")
generator = generator.manual_seed(38)
img = pipe("An astronaut riding a dinossaur", generator=generator)[0]
img[0]
```

### System Info

- `diffusers` version: 0.12.1
- Platform: Linux-5.10.147+-x86_64-with-glibc2.29
- Python version: 3.8.10
- PyTorch version (GPU?): 1.13.1+cu116 (True)
- Huggingface_hub version: 0.12.0
- Transformers version: 4.26.0
- Accelerate version: not installed
- xFormers version: not installed
- Using GPU in script?: T4

**Comment by patrickvonplaten** [comment]
Hey @jorgemcgomes,

Thanks a lot for opening such a clean issue. I took a deeper look into things and I'm quite certain that the current implementation is **correct**. 

Maybe you can confirm the following:

1. The original stability AI code bases uses the open_clip tokenizer, see:
- https://github.com/Stability-AI/stablediffusion/blob/d55bcd4d31d0316fcbdf552f2fd2628fdc812500/configs/stable-diffusion/v2-inference.yaml#L64
2. More specifically they just call the function `open_clip.tokenize(...)` from the FrozenOpenclip class as shown here: 
- https://github.com/Stability-AI/stablediffusion/blob/d55bcd4d31d0316fcbdf552f2fd2628fdc812500/ldm/modules/encoders/modules.py#LL169C9-L169C42

=> Now let's see what this does:

```
pip install open_clip_torch
```

and then 
```py
import open_clip
print(open_clip("test tokenization")
```
gives:
```
tensor([[49406,  1628, 32634, 10847, 49407,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0,     0,     0,     0,
             0,     0,     0,     0,     0,     0,     0]])
```

therefore matching what we currently have in v2. 

So I think we should be good no? 

Also cc @patil-suraj

**Comment by jorgemcgomes** [comment]
Thanks for looking into this. Your argument seems solid.

This is still very odd though. And it means that the tokenizer available at https://huggingface.co/laion/CLIP-ViT-H-14-laion2B-s32B-b79K doesn't tokenize the same as the OpenCLIP implementation. But I guess it doesn't make difference in most use cases because CLIP typically uses the attention mask?

I ran into this issue while trying to use SD2.1 with the pipeline https://github.com/huggingface/diffusers/blob/main/examples/community/lpw_stable_diffusion.py .
The strategy used there to handle long text inputs (split in chunks) doesn't work this tokenization strategy, where the padding token is different from the EOS token.

https://github.com/huggingface/diffusers/blob/58c416ab0c8859989da6cd225b31831b6a425b15/examples/community/lpw_stable_diffusion.py#L207

**Comment by patrickvonplaten** [comment]
I agree it's odd! @jorgemcgomes would you say you consistently get better results when using the v1 tokenizer for v2? 

In that case, we should prob contact stability AI

**Comment by jorgemcgomes** [comment]
I didn't test it exhaustively, but in the few experiments I did the results were inconclusive. The results were often quite different (sometimes much more different than the examples I posted), but there wasn't a clear winner.

**Comment by patil-suraj** [comment]
Hey @jorgemcgomes, Thanks for the issue. As Patrick said, the diffusers implem matches 1:1 with the original SD2.0 repo.
I have also observed that sometimes using eos can give better results, but can't say that it works in all situations.
